src/core/network.c

```diff
@@ -392,95 +392,68 @@ int net_getsockname(GIOChannel *handle, IPADDR *addr, int *port)
 	return 0;
 }
 
-/* Get IP addresses for host, both IPv4 and IPv6 if possible.
-   If ip->family is 0, the address wasn't found.
-   Returns 0 = ok, others = error code for net_gethosterror() */
-int net_gethostbyname(const char *addr, IPADDR *ip4, IPADDR *ip6)
+/* Get IP addresses for host, both IPv4 and IPv6 if possible. */
+RESOLVED_IP_REC *net_gethostbyname(const char *addr, GResolverNameLookupFlags flags)
 {
-	union sockaddr_union *so;
-	struct addrinfo hints, *ai, *ailist;
-	int ret, count_v4, count_v6, use_v4, use_v6;
+	/* GList<GInetAddress> */
+	GList *ailist;
+	GError *error;
+	GResolver *resolver;
+	RESOLVED_IP_REC *iprec;
 
 #ifdef HAVE_CAPSICUM
 	if (capsicum_enabled())
-		return (capsicum_net_gethostbyname(addr, ip4, ip6));
+		return (capsicum_net_gethostbyname(addr, flags));
 #endif
 
-	g_return_val_if_fail(addr != NULL, -1);
+	g_return_val_if_fail(addr != NULL, NULL);
 
-	memset(ip4, 0, sizeof(IPADDR));
-	memset(ip6, 0, sizeof(IPADDR));
-
-	memset(&hints, 0, sizeof(struct addrinfo));
-	hints.ai_socktype = SOCK_STREAM;
-	hints.ai_flags = AI_ADDRCONFIG;
-
-	/* save error to host_error for later use */
-	ret = getaddrinfo(addr, NULL, &hints, &ailist);
-	if (ret != 0)
-		return ret;
-
-	/* count IPs */
-        count_v4 = count_v6 = 0;
-	for (ai = ailist; ai != NULL; ai = ai->ai_next) {
-		if (ai->ai_family == AF_INET)
-			count_v4++;
-		else if (ai->ai_family == AF_INET6)
-			count_v6++;
+	error = NULL;
+	resolver = g_resolver_get_default();
+	ailist = g_resolver_lookup_by_name_with_flags(resolver, addr, flags, NULL, &error);
+	iprec = g_new0(RESOLVED_IP_REC, 1);
+	if (error != NULL) {
+		iprec->error = error;
+	} else {
+		iprec->ailist = ailist;
 	}
+	g_object_unref(resolver);
 
-	if (count_v4 == 0 && count_v6 == 0)
-		return EAI_NONAME; /* shouldn't happen? */
-
-	/* if there are multiple addresses, return random one */
-	use_v4 = count_v4 <= 1 ? 0 : rand() % count_v4;
-	use_v6 = count_v6 <= 1 ? 0 : rand() % count_v6;
-
-	count_v4 = count_v6 = 0;
-	for (ai = ailist; ai != NULL; ai = ai->ai_next) {
-		so = (union sockaddr_union *) ai->ai_addr;
-
-		if (ai->ai_family == AF_INET) {
-			if (use_v4 == count_v4)
-				sin_get_ip(so, ip4);
-                        count_v4++;
-		} else if (ai->ai_family == AF_INET6) {
-			if (use_v6 == count_v6)
-				sin_get_ip(so, ip6);
-			count_v6++;
-		}
-	}
-	freeaddrinfo(ailist);
-	return 0;
+	return iprec;
 }
 
-/* Get name for host, *name should be g_free()'d unless it's NULL.
-   Return values are the same as with net_gethostbyname() */
-int net_gethostbyaddr(IPADDR *ip, char **name)
+int net_gethostbyname_first_ips(const char *addr, GResolverNameLookupFlags flags, IPADDR *ip4,
```

Comment:
This function is throwing away important information:
1. the relative order between IPv4 and IPv6 (which one comes first)
2. the second and further addresses of each family.

*Most* DNS resolvers will change the order of the results if you look up the same thing again, but that's not required. So by throwing away the second and later results, we could end up stuck trying to connect to a server that fails to respond. This is a pre-existing problem.